FlashBuilder refactor

Dirbaio opened the pull request:
This PR refactors the FlashBuilder code, to simplify it.

- FlashLoader owns all the data in a `BTreeMap<u32, Vec<u8>>`
- FlashLoader has a `data_in_range` function allowing iterating data inside a range. Data blocks that are not fully contained inside the range are cut, so calling code can assume all the returned blocks are fully contained, drastically simplifying the code.
- FlashBuilder iterates over all sectors and pages, requesting data contained in them using `data_in_range`.

The new FlashBuilder passes all the existing tests with no modifications (ignoring order, because the old code returned unsorted data in one test and the new code always returns sorted data)

TODO: 
- [x] Error when the FlashLoader has data not in any MemoryRegion. (currently it's silently ignored)

Yatekii reviewed (commented):
> [ ] Error when the FlashLoader has data not in any MemoryRegion. (currently it's silently ignored)

I think data that is not in any memory region should not be able to enter the flash Builder in the first place.

Dirbaio commented on probe-rs/src/flashing/download.rs:
```diff
@@ -178,67 +166,6 @@ impl std::fmt::Debug for ExtractedFlashData<'_> {
     }
 }
 
-impl<'data> ExtractedFlashData<'data> {
```
With the changes it seems to be completely unused now, isn't it?

Tiwalun replied:
I think the plan for this was to use in the returned error, i.e. in the `FlashError::NoSuitableNvm` variant. The goal was to show the offending ELF section in the error message, to help with debugging. It seems that we never got around to actually use it, though.

Dirbaio commented:
- Added error checking for adding data in regions not in the memory map.
- Moved the `data` BTreeMap from `FlashLoader` to `FlashBuilder`. This is closer to how the code worked before and makes more sense for tests, for example.

This is now fully done :)

Yatekii commented on probe-rs/src/flashing/flash_algorithm.rs:
```diff
@@ -91,6 +91,58 @@ impl FlashAlgorithm {
+        })
+    }
+
+    /// Iterate over all the pages of the flash
```
```suggestion
    /// Iterate over all the pages of the flash.
```

Yatekii commented on probe-rs/src/flashing/builder.rs:
```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
+        // Check the new data doesn't overlap to the right
+        if let Some((&next_addr, _)) = self.data.range(address..).next() {
+            if address + (data.len() as u32) > next_addr {
+                panic!("data overlaps to the right");
```
Since you made this function public I am not okay with the panic anymore. It's not okay for the library to explode when there is user-error.

Dirbaio replied:
Changed to an error.

Yatekii commented on probe-rs/src/flashing/builder.rs:
```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
-                        address
-                    );
-                }
+        // Check the new data doesn't overlap to the right
```
```suggestion
        // Check the new data doesn't overlap to the right.
```

Yatekii commented on probe-rs/src/flashing/builder.rs:
```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
+            let prev_end = prev_addr + (prev_data.len() as u32);
+
+            if prev_end > address {
+                panic!("data overlaps to the left");
```
Same as with the panic above.

Dirbaio replied:
Changed to an error.

Yatekii commented on probe-rs/src/flashing/builder.rs:
```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
-                // If we made it until here, it is safe to insert the block.
-                self.data_blocks
-                    .insert(position, FlashDataBlock::new(address, data.to_vec()))
+        // Check the new data doesn't overlap to the left
```
```suggestion
        // Check the new data doesn't overlap to the left.
```

Yatekii commented on probe-rs/src/flashing/builder.rs:
```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
+    ///
+    /// Data is returned in ascending address order, and is guaranteed not to overlap.
+    /// If a staged chunk is not fully contained in the range, only the contained part is
+    /// returned. ie it's guaranteed that `start <= addr && addr+len <= end`
```
```suggestion
    /// returned. ie it's guaranteed that `start <= addr && addr + len <= end` upholds.
```

Yatekii commented on probe-rs/src/flashing/builder.rs:
```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
+                let mut addr = addr;
+                let mut data = &data[..];
+
+                // Cut chunk from the left if it starts before `start`
```
```suggestion
                // Cut chunk from the left if it starts before `start`.
```

Yatekii commented on probe-rs/src/flashing/builder.rs:
```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
-                        pages.len() - 1,
-                    );
-                }
+            // Ignore if no data.
```
```suggestion
            // Ignore if neither the sector nor the page contain any data.
```

Yatekii commented on probe-rs/src/flashing/builder.rs:
```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
+            let page_end = info.base_address + info.size;
+            let range = info.base_address..page_end;
+
+            // Ignore if outside the NvmRegion.
```
```suggestion
            // Ignore the page if it's outside the NvmRegion.
```

Yatekii commented on probe-rs/src/flashing/builder.rs:
```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
+            let range = info.base_address..page_end;
+
+            // Ignore if outside the NvmRegion.
+            if !region.range.contains_range(&range.clone()) {
```
How can a page ever not be contained in a memory range?

Dirbaio replied:
We're iterating all the pages covered by the flash algorithm, which may be more than the NvmRegion if the algo manages multiple NvmRegions.

Yatekii replied:
Ah, I see! Thanks!

Yatekii commented on probe-rs/src/flashing/builder.rs:
```diff
@@ -194,313 +155,213 @@ impl FlashDataBlockSpan {
 
-                // Make sure we advance the block offset by the amount we just wrote.
-                block_offset += size;
+            // Fill hole between the last datablock (or page start if no blocks) and page end.
```
```suggestion
            // Fill the hole between the last data block (or page start if there is no blocks) and page end.
```

Yatekii reviewed (changes requested):
Thanks a lot for fixing this and cleaning it up a lot! It looks much better now I believe :)

I have found a few panics/asserts I am not okay with. Maybe I understood it wrong, but I am basically okay if they are not in public API. If they are I do not consider it okay.
Most of them should be easily fixable I think, otherwise we can work something out.
If I understood it wrong, please let me know :)

I added a lot of full stops and some wording in comments as that was the only things I could really complain about :P Feel free to change those or leave them as is.

Dirbaio commented on probe-rs/src/flashing/flash_algorithm.rs:
```diff
@@ -91,6 +91,58 @@ impl FlashAlgorithm {
+        let props = &self.flash_properties;
+
+        assert!(!props.sectors.is_empty());
+        assert!(props.sectors[0].address == 0);
```
These addresses are relative to the flash start, so the first sector should indeed start at 0. If it starts at eg 0x1000, the sector size of 0x0-0x1000 isn't known

Yatekii reviewed (approved):
LGTM, thanks!

bors r+
